src/realm/sync/noinst/client_reset.cpp

```diff
@@ -426,110 +446,203 @@ void remove_pending_client_resets(Transaction& wt)
+        REALM_ASSERT(recovery_col);
+        REALM_ASSERT(action_col);
+        REALM_ASSERT(code_col);
+        REALM_ASSERT(msg_col);
+        pending.recovery_allowed = first.get<bool>(recovery_col);
+        pending.action = to_reset_action(first.get<int64_t>(action_col));
+        auto code = first.get<int64_t>(code_col);
+        if (code > 0) {
+            pending.error = Status(static_cast<ErrorCodes::Error>(code), first.get<StringData>(msg_col));
+        }
     }
     else {
-        throw ClientResetFailed(
-            util::format("Unsupported client reset metadata type: %1 from %2", type, pending.time));
+        // Provide default action and recovery values if table version is 1
+        pending.recovery_allowed = true;
+        pending.action = PendingReset::Action::ClientReset;
     }
     return pending;
 }
 
-void track_reset(Transaction& wt, ClientResyncMode mode)
+static TableRef get_or_create_client_reset_table(Transaction& wt)
 {
-    REALM_ASSERT(mode != ClientResyncMode::Manual);
     TableRef table = wt.get_table(s_meta_reset_table_name);
-    ColKey version_col, timestamp_col, type_col;
     if (!table) {
+        // Creating new table
         table = wt.add_table_with_primary_key(s_meta_reset_table_name, type_ObjectId, s_pk_col_name);
         REALM_ASSERT(table);
-        version_col = table->add_column(type_Int, s_version_column_name);
-        timestamp_col = table->add_column(type_Timestamp, s_timestamp_col_name);
-        type_col = table->add_column(type_Int, s_reset_type_col_name);
-    }
-    else {
-        version_col = table->get_column_key(s_version_column_name);
-        timestamp_col = table->get_column_key(s_timestamp_col_name);
-        type_col = table->get_column_key(s_reset_type_col_name);
-    }
-    REALM_ASSERT(version_col);
-    REALM_ASSERT(timestamp_col);
-    REALM_ASSERT(type_col);
-    int64_t mode_val = 0; // Discard
-    if (mode == ClientResyncMode::Recover || mode == ClientResyncMode::RecoverOrDiscard) {
-        mode_val = 1; // Recover
+        table->add_column(type_Int, s_version_column_name);
+        table->add_column(type_Timestamp, s_timestamp_col_name);
+        table->add_column(type_Int, s_reset_mode_col_name);
+        table->add_column(type_Bool, s_reset_recovery_col_name);
+        table->add_column(type_Int, s_reset_action_col_name);
+        table->add_column(type_Int, s_reset_error_code_col_name);
+        table->add_column(type_String, s_reset_error_msg_col_name);
     }
+    return table;
+}
 
-    if (table->size() > 1) {
+void track_reset(Transaction& wt, ClientResyncMode mode, bool recovery_allowed, PendingReset::Action action,
+                 const std::optional<Status>& error)
+{
+    REALM_ASSERT(mode != ClientResyncMode::Manual);
+    TableRef table = get_or_create_client_reset_table(wt);
+    REALM_ASSERT(table);
+    // Even if the table is being updated to V2, an existing entry will still throw an exception
+    size_t table_size = table->size();
```

Comment:
I think 583/584 can just be `if(table->size() > 0)`. it doesn't look like we use the table_size variable anywhere else.